pkg/backend/endpoint_resolver.go

```diff
@@ -121,6 +129,90 @@ func (r *defaultEndpointResolver) ResolvePodEndpoints(ctx context.Context, svcKe
 	return endpoints, containsPotentialReadyEndpoints, nil
 }
 
+func (r *defaultEndpointResolver) ResolvePodEndpointsFromSlices(ctx context.Context, svcKey types.NamespacedName, port intstr.IntOrString,
+	opts ...EndpointResolveOption) ([]PodEndpoint, bool, error) {
+	resolveOpts := defaultEndpointResolveOptions()
+	resolveOpts.ApplyOptions(opts)
+
+	svc, svcPort, err := r.findServiceAndServicePort(ctx, svcKey, port)
+	if err != nil {
+		return nil, false, err
+	}
+
+	epSlicesList := &discv1.EndpointSliceList{}
+	if err := r.k8sClient.List(ctx, epSlicesList,
+		client.InNamespace(svcKey.Namespace),
+		client.MatchingLabels{"kubernetes.io/service-name": svcKey.Name}); err != nil {
+		return nil, false, err
+	}
+	if len(epSlicesList.Items) == 0 {
+		return nil, false, fmt.Errorf("%w: endpointslices for \"%v\" not found", ErrNotFound, svcKey.Name)
+	}
+
+	containsPotentialReadyEndpoints := false
+	var endpoints []PodEndpoint
+	used_addr := make(sets.String)
+	for _, epSlice := range epSlicesList.Items {
+		// process epSlice
+		for _, epPort := range epSlice.Ports {
+			// servicePort.Name is optional if there is only one port
+			if svcPort.Name != "" && svcPort.Name != *epPort.Name {
+				continue
+			}
+			for _, ep := range epSlice.Endpoints {
+				if ep.Conditions.Ready == nil || *ep.Conditions.Ready {
+					for _, epAddr := range ep.Addresses {
+						if used_addr.Has(epAddr) {
+							continue
+						}
+						if ep.TargetRef == nil || ep.TargetRef.Kind != "Pod" {
+							continue
+						}
+						pod, exists, err := r.findPodByReference(ctx, svc.Namespace, *ep.TargetRef)
+						if err != nil {
+							return nil, false, err
+						}
+						if !exists {
+							return nil, false, errors.New("couldn't find podInfo for ready endpoint")
+						}
+						used_addr.Insert(epAddr)
+						endpoints = append(endpoints, buildPodEndpointFromSlice(pod, epAddr, epPort))
+					}
+				}
+				if len(resolveOpts.PodReadinessGates) != 0 {
+					for _, epAddr := range ep.Addresses {
+						if used_addr.Has(epAddr) {
+							continue
+						}
+						if ep.TargetRef == nil || ep.TargetRef.Kind != "Pod" {
+							continue
+						}
+						pod, exists, err := r.findPodByReference(ctx, svc.Namespace, *ep.TargetRef)
+						if err != nil {
+							return nil, false, err
+						}
+						if !exists {
+							containsPotentialReadyEndpoints = true
+							continue
+						}
+						if !pod.HasAnyOfReadinessGates(resolveOpts.PodReadinessGates) {
+							continue
+						}
+						if !pod.IsContainersReady() {
+							containsPotentialReadyEndpoints = true
+							continue
+						}
+						used_addr.Insert(epAddr)
+						endpoints = append(endpoints, buildPodEndpointFromSlice(pod, epAddr, epPort))
+					}
```

Comment:
since you already loop through the ep.Addresses in line 64, IMO, we can consolidate these two loops and add the the logic to check whether endpoint ready status is true.